README.rst
wpsmon
======

Setup
-----
If you want to use an external (usb?) wifi dongle as your monitor device, here's
a guide on how to set it up.

First take note of the wifi channel you want to scan. This can be done using
the following command:

    sudo iwlist wlan0 scan | egrep 'Address:|ESSID:|Channel:'

On debian you may need to install the package wireless-tools to have the iwlist
tool available.

    sudo aptitude update && aptitude install wireless-tools

The output from this command will resemble this:

    ...
    Cell 07 - Address: C0:4A:00:2D:15:C0
            Channel:36
            ESSID:"steinwurf-test-5g"
    Cell 08 - Address: 28:10:7B:8E:2A:F0
            Channel:1
            ESSID:"ChocolateCloud"
    Cell 09 - Address: B4:75:0E:D0:B5:B9
            Channel:11
            ESSID:"ccdc"
    ...

From this it can be seen that there's a cell for each ssid, e.g., the
steinwurf-test-5g network is on channel 36.

Now for setting up the monitor interface...

To make sure the network manager is not interfering with our business, we need
to add the following line to `/etc/network/interfaces` (from
http://superuser.com/q/9720):

    iface mon0 inet manual

After adding this line, restart the network manager service to make sure it
picks up the changes:

    sudo service network-manager stop
    sudo service network-manager start

To add the mon0 monitor interface, we do the following
(from https://sandilands.info/sgordon/capturing-wifi-in-monitor-mode-with-iw):

To see the name of the physical adapter we want to add the monitor interface on,
we use the following command:

    iw dev

Add the monitor interface, where phy0 is the name found in the previous step:

    sudo iw phy phy0 interface add mon0 type monitor

We will capture with the mon0 interface, so you can delete the normal wlan0
interface:

    sudo iw dev wlan0 del

Now enable the mon0 interface using ifconfig:

    sudo ifconfig mon0 up

If you get the following error:

    SIOCSIFFLAGS: Cannot allocate memory

You may need to install the non-free ralink driver:

    sudo aptitude update && aptitude install firmware-ralink

Now we need to set the channel to monitor:

    sudo iw dev mon0 set channel 36

Here we are monitoring the steinwurf-test-5g on channel 36.


Install
-------

To use the script you need to install a few dependencies:

    sudo aptitude install libpcap-dev

and a few python packages, let's make a virtual env:

    virtualenv wpsmon
    source wpsmon/bin/activate
    pip install -r ./requirements.txt

Run
---

When you have set everything up you can run the tool using the following
approach.

First, if you have the virtual env enabled, exit it by using this commmand:

    deactivate

Login as root and activate the virtualenv:

    su
    source wpsmon/bin/activate

You can now start the tool:

    ./wpsmon.py mon0

Introduction
------------
wpsmon is a tool to monitor the devices connected to a certain wifi. It can give
you information about which devices are in sleep mode and which are in active
mode.

The UI is a table with the following columns:

* bssid: the mac address of the device.
* ps: if 1 the device is in power save mode, if 0 the device is in active mode.
* frames: total number of data frames seen from device.
* slept: total number of times the devices have been in power save mode.
* tout: the time the device have been in power save mode (in milliseconds).
* tmax: the maximum time the device have been in power save mode
  (in milliseconds).
* alias/ip: the alias (human friendly name) or IP of the device (the alias can
  be specified using the alias file).

Each row is a device and the color of the text describes the status of the
device:
* green: device in active mode.
* red: device in power save mode.
* gray: device is stale i.e. we have not heard from device in SEC seconds
  (where SEC can be adjusted using the -s, --stale-time argument, default is 30)

If a device have been silent for SEC seconds, it will be removed from the list
(SEC can be adjusted using the -d, --dead-time argument, default is 60).

Sources
-------
802.11-2012 Standard:
http://standards.ieee.org/getieee802/download/802.11-2012.pdf
8.2.4.1.7 Power Management field


NEWS.rst
News for wpsmon
===============

This file lists the major changes between versions. For a more
detailed list of every change, see the Git log.

Latest
------
* tbd


wpsmon.py
#!/usr/bin/env python2

"""
Dump 802.11 power-save status

Add a monitor interface and specify channel before use:
  iw phy <phy> interface add mon0 type monitor
  iw mon0 set channel <channel> HT20

Replace <phy> and <channel> with the correct values.

Alternative setup:
  ifconfig <wlan device> down
  iwconfig <wlan device> mode monitor
  ifconfig <wlan device> up
  iw <wlan device> set channel <channel> HT20

  run wpsmon.py with <wlan device> as interface
"""

from __future__ import print_function

import sys
import os
import time
import datetime
import argparse
import socket
import re
import curses
import subprocess
import dpkt
import pcapy


class wpsmon():

    def __init__(self, interface, timeout_ms=250):

        self.captured = 0
        self.stations = {}
        self.alias = {}
        self.ips = {}

        self.stale_time = 0
        self.dead_time = 0

        self.interface = interface
        self.only_alias = False

        self.prog = sys.argv[0].replace('./', '')

        # Setup capture
        self.pc = pcapy.open_live(interface, 65536, 1, timeout_ms)

    def set_screen(self, screen):
        self.screen = screen

    def set_stale_time(self, stale_time):
        self.stale_time = stale_time

    def set_dead_time(self, dead_time):
        self.dead_time = dead_time

    def set_only_alias(self, only_alias):
        self.only_alias = only_alias

    def update_ip_list(self):
        output = subprocess.check_output(['ip', 'neighbor', 'show'])
        ip_neigh = str(output).split('\n')

        for entry in ip_neigh:
            try:
                m = re.split('[\s]+', entry)
                ip = m[0].strip()
                lladdr = m[4].strip().lower()
                self.ips[lladdr] = ip
            except:
                pass

    def mac_string(self, mac):
        return ':'.join('{0:02X}'.format(ord(b)) for b in mac)

    def next(self):
        (header, packet) = self.pc.next()
        if header and packet:
            self.parse_packet(header, packet)

    def update_timeout(self):
        now = time.time()
        for station in self.stations.values():
            age = now - station['last']
            if self.stale_time > 0 and age > self.stale_time:
                station['stale'] = True
            if self.dead_time > 0 and age > self.dead_time:
                self.stations.pop(station['src'])

    def update_window(self):
        self.screen.clear()

        # Update stale nodes
        self.update_timeout()

        # Update MAC to IP table
        self.update_ip_list()

        nodes = len(self.stations)

        now = datetime.datetime.now().strftime('%Y-%m-%d %H:%M')

        top = '\n [ {0} ][ captured {1} frames from {2} nodes ][ {3} ]\n\n'
        self.screen.addstr(top.format(
            self.prog, self.captured, nodes, now))

        header = ' {bssid:18s} {ps:3s} {frames:7s} {slept:5s}' \
                 '{tout:>7s} {tmax:>7s}  {alias}\n\n'
        self.screen.addstr(header.format(**
                           {'bssid': 'bssid',
                            'ps': 'ps',
                            'frames': 'frames',
                            'slept': 'slept',
                            'tout': 'tout',
                            'tmax': 'tmax',
                            'alias': 'alias/ip'}))

        # Sort stations according to creation time
        sorted_stations = sorted(
            self.stations.values(),
            key=lambda s: int(s['created'] * 1000))

        # Get window dimensions
        maxy, maxx = self.screen.getmaxyx()

        shown = 0
        for station in sorted_stations:
            # Break if we cant fit more clients on the screen
            y, x = self.screen.getyx()
            if y >= maxy - 3:
                overflow = nodes - shown
                self.screen.addstr(" {0} nodes not shown...".format(overflow))
                break
            shown += 1

            # Continue if only showing aliased nodes
            if self.only_alias and not station['alias']:
                continue

            fmt = ' {src:18s} {ps:<3d} {frames:<7d} {slept:<5d}'\
                  '{tout:>7.1f} {tmax:>7.1f}  {alias}{ip}\n'
            text = fmt.format(**station)
            if station['stale']:
                color = curses.color_pair(3) | curses.A_BOLD
            elif station['ps']:
                color = curses.color_pair(1) | curses.A_BOLD
            else:
                color = curses.color_pair(2)
            try:
                self.screen.addstr(text, color)
            except:
                # Too many clients to show
                break

        # Show help text
        shortcuts = "q: quit | r: reset counters | R: reset nodes"
        self.screen.addstr(maxy-1, 1, shortcuts)

        self.screen.refresh()

    def add_alias(self, host, name):
            self.alias[host.lower()] = name

    def reset_counters(self):
        self.captured = 0
        for station in self.stations.values():
            station['frames'] = 0
            station['slept'] = 0
            station['tout'] = 0
            station['tmax'] = 0

    def reset_nodes(self):
        self.stations = {}
        self.reset_counters()

    def parse_packet(self, header, packet):

        self.captured += 1

        # Read radiotap header length
        try:
            # todo let's output the errors somewhere.
            tap = dpkt.radiotap.Radiotap(packet)
        except dpkt.Error:
            return
        except KeyError:
            return

        tap_len = socket.ntohs(tap.length)

        # Parse IEEE80211 header
        try:
            wlan = dpkt.ieee80211.IEEE80211(packet[tap_len:])
        except dpkt.Error:
            return

        # Currently we only care about data frames
        if wlan.type is not dpkt.ieee80211.DATA_TYPE:
            return

        ps = wlan.pwr_mgt
        src = self.mac_string(wlan.data_frame.src).lower()

        # Lookup station
        station = self.stations.get(src)

        # Get current time
        now = time.time()

        # New station
        if not station:
            self.stations[src] = {}
            station = self.stations[src]
            station['src'] = src
            station['alias'] = self.alias.get(src, '')
            station['ip'] = ''
            station['created'] = now
            station['frames'] = 0
            station['tout'] = 0
            station['tmax'] = 0
            station['slept'] = 0

        # Detect if a station is going to sleep
        old_ps = station.get('ps', 0)
        station['ps'] = ps
        going_to_ps = ps and not old_ps

        # Count number of sleeps
        if going_to_ps:
            station['slept'] += 1

        # Calculate timeout if going to PS
        if 'last' in station and going_to_ps:
            diff_ms = (now - station['last']) * 1000
            station['tout'] = diff_ms

            if diff_ms > station['tmax']:
                station['tmax'] = diff_ms

        # Log last updated time
        station['last'] = now

        # Increment packet frame count
        station['frames'] += 1

        # Try to set IP if empty
        if station['ip'] == '':
            station['ip'] = self.ips.get(src, '')
            if station['ip'] != '' and station['alias'] != '':
                station['ip'] = ' (' + station['ip'] + ')'

        # Station is not stale
        station['stale'] = False

def alias_split(alias):
    match = re.match('(..:..:..:..:..:..)=(.*)', alias, flags=re.IGNORECASE)
    if not match:
        raise RuntimeError('Failed to parse alias: ' + alias)
    return match.group(1), match.group(2)

def alias_type(alias):
    try:
        host, name = alias_split(alias)
    except Exception as e:
        raise argparse.ArgumentTypeError(e)
    return (host, name)

def main():
    formatter = argparse.RawTextHelpFormatter

    parser = argparse.ArgumentParser(description=__doc__,
                                     formatter_class=formatter)

    parser.add_argument('interface', help='interface to sniff')
    parser.add_argument('-a', '--alias', metavar='<mac=name>',
                        action='append', type=alias_type,
                        help='alias mac with name')
    parser.add_argument('-f', '--alias-file', metavar='<file>',
                        help='read aliases from file',
                        default='steinwurf_alias.txt')
    parser.add_argument('-A', '--only-alias', action='store_true',
                        help='only show aliased nodes')
    parser.add_argument('-s', '--stale-time',
                        type=int, default=30, metavar='<sec>',
                        help='consider node stale after SEC seconds')
    parser.add_argument('-d', '--dead-time',
                        type=int, default=60, metavar='<sec>',
                        help='consider node dead after SEC seconds')

    args = parser.parse_args()

    # Create monitor object
    try:
        mon = wpsmon(args.interface)
    except Exception as e:
        print("Failed to open capture: " + str(e))
        sys.exit(os.EX_NOPERM)

    # Setup timeouts
    mon.set_stale_time(args.stale_time)
    mon.set_dead_time(args.dead_time)

    # Map aliases from command line
    if args.alias is not None:
        for a in args.alias:
            host, name = a
            mon.add_alias(host, name)

    # Map aliases from file
    if args.alias_file is not None:
        with open(args.alias_file) as f:
            for line in f:
                # Skip comments and empty lines
                if re.match('^\s*(#.*)?$', line):
                    continue
                host, name = alias_split(line)
                mon.add_alias(host, name)

    mon.set_only_alias(args.only_alias)

    # Setup curses
    stdscr = curses.initscr()
    curses.noecho()
    curses.cbreak()
    curses.curs_set(0)
    stdscr.nodelay(1)

    # Setup colors
    curses.start_color()
    curses.use_default_colors()
    curses.init_pair(1, curses.COLOR_RED, -1)
    curses.init_pair(2, curses.COLOR_GREEN, -1)
    curses.init_pair(3, curses.COLOR_BLACK, -1)

    # Setup screen
    mon.set_screen(stdscr)

    last_update = 0
    while True:
        now = time.time()
        if now > last_update + 0.1:
            mon.update_window()
            last_update = now
        try:
            mon.next()
            ch = stdscr.getch()
            if ch == ord('q'):
                break
            if ch == ord('r'):
                mon.reset_counters()
            if ch == ord('R'):
                mon.reset_counters()
                mon.reset_nodes()
        except KeyboardInterrupt:
            break

    # Cleanup curses
    curses.nocbreak()
    curses.echo()
    curses.curs_set(1)
    curses.endwin()

if __name__ == '__main__':
    main()
